pricegraph/src/orderbook/iter.rs

```diff
@@ -0,0 +1,203 @@
+//! Module containing transitive order iterator for an orderbook.
+
+use crate::{
+    encoding::TokenPair,
+    graph::path::Path,
+    orderbook::{self, Flow, Orderbook, OverlapError, ReducedOrderbook},
+};
+use petgraph::graph::NodeIndex;
+use std::{iter::FusedIterator, mem};
+
+/// An iterator over all transitive orders over a token pair for an orderbook,
+/// ordered from lowest limit price to highest.
+pub struct TransitiveOrders {
+    orderbook: Orderbook,
+    pair: Option<(NodeIndex, NodeIndex)>,
+    state: State,
+}
+
+/// Internal iterator state.
+enum State {
+    /// The next transitive order has already been pre-computed and is ready to
+    /// be returned.
+    NextOrder(Option<(Path<NodeIndex>, Flow)>),
+    /// The previous order needs to be filled before continuing.
+    FillPreviousOrder(Path<NodeIndex>, Flow),
+    /// The next transitive order needs to be found, and no previous orders
+    /// need to be filled.
+    Search,
+}
+
+impl TransitiveOrders {
+    /// Creates an empty transitive orderbook iterator.
+    fn empty(orderbook: Orderbook) -> Self {
+        Self {
+            orderbook,
+            pair: None,
+            state: State::Search,
+        }
+    }
+
+    /// Creates a new transitive orderbook iterator.
+    pub fn new(orderbook: Orderbook, pair: TokenPair) -> Result<Self, OverlapError> {
+        let (buy, sell) = if orderbook.is_token_pair_valid(pair) {
+            (
+                orderbook::node_index(pair.buy),
+                orderbook::node_index(pair.sell),
+            )
+        } else {
+            return Ok(Self::empty(orderbook));
+        };
+
+        // NOTE: We need to check that the orderbook is not overlapping in the
+        // subgraph containing the token pair we care about, so we find the
+        // first transitive order and reuse the result in the first call to
+        // `next`.
+        let next_order = orderbook.find_path_and_flow(buy, sell)?;
```

Comment:
Then we could also return iter::empty instead of needing to have pair optional to signal emptiness.

Reply:
Why is this logic different than in the reduced case? Is there some other reason than returning Result vs Self (which makes sense)? Still wondering if more logic could be shared between the two?

Reply:
> if this returned impl Iterator we could use std::iter::chain and remove the NextOrder state [..] Then we could also return iter::empty instead of needing to have pair optional

Even if we return `impl Iterator` we can only return one type, alternatively we could `Box` it, but I'm not sure how I feel about it.

> Is there some other reason than returning Result vs Self (which makes sense)?

It was a silly optimization that doesn't really make sense: for reduced orderbooks, we don't need to find a path to make sure there are no negative cycles, so we can do it lazily in the first `next` call. Since `next` should get called at least once always, its not really a useful optimization and just having one path makes things easier to read.

Reply:
I forgot that we would have to Box. Agree that in that case it's not great.